### ISSUE TITLE
Unable to replace Categorical RandomVariable using a data dict during inference

### ISSUE THREAD

**Issue by jhaberstro** [blei-lab/edward#407]
The `Categorical` class expects that the parameter to `log_prob` be an integer type. When replacing a `Categorical` RV during inference with integer data:

```
data = [1, 2, 3]
x = Categorical(logits=..., validate_args=True)
...
inference = ed.KLqp({...},
                    {x: tf.constant(data)})
inference.run()
```

the following assert fires:

```
  File "lda.py", line 60, in <module>
    inference.run()
  File "/Users/jhaberstro/anaconda/envs/datascience/lib/python3.5/site-packages/edward/inferences/inference.py", line 211, in run
    self.initialize(*args, **kwargs)
  File "/Users/jhaberstro/anaconda/envs/datascience/lib/python3.5/site-packages/edward/inferences/klqp.py", line 66, in initialize
    return super(KLqp, self).initialize(*args, **kwargs)
  File "/Users/jhaberstro/anaconda/envs/datascience/lib/python3.5/site-packages/edward/inferences/variational_inference.py", line 70, in initialize
    self.loss, grads_and_vars = self.build_loss_and_gradients(var_list)
  File "/Users/jhaberstro/anaconda/envs/datascience/lib/python3.5/site-packages/edward/inferences/klqp.py", line 117, in build_loss_and_gradients
    return build_score_loss_and_gradients(self, var_list)
  File "/Users/jhaberstro/anaconda/envs/datascience/lib/python3.5/site-packages/edward/inferences/klqp.py", line 567, in build_score_loss_and_gradients
    x_log_prob = tf.reduce_sum(x_copy.log_prob(dict_swap[x]))
  File "/Users/jhaberstro/anaconda/envs/datascience/lib/python3.5/site-packages/tensorflow/contrib/distributions/python/ops/distribution.py", line 437, in log_prob
    return self._log_prob(value)
  File "/Users/jhaberstro/anaconda/envs/datascience/lib/python3.5/site-packages/tensorflow/contrib/distributions/python/ops/categorical.py", line 143, in _log_prob
    return -nn_ops.sparse_softmax_cross_entropy_with_logits(logits, k)
  File "/Users/jhaberstro/anaconda/envs/datascience/lib/python3.5/site-packages/tensorflow/python/ops/nn_ops.py", line 778, in sparse_softmax_cross_entropy_with_logits
    precise_logits, labels, name=name)
  File "/Users/jhaberstro/anaconda/envs/datascience/lib/python3.5/site-packages/tensorflow/python/ops/gen_nn_ops.py", line 1857, in _sparse_softmax_cross_entropy_with_logits
    features=features, labels=labels, name=name)
  File "/Users/jhaberstro/anaconda/envs/datascience/lib/python3.5/site-packages/tensorflow/python/framework/op_def_library.py", line 573, in apply_op
    _Attr(op_def, input_arg.type_attr))
  File "/Users/jhaberstro/anaconda/envs/datascience/lib/python3.5/site-packages/tensorflow/python/framework/op_def_library.py", line 60, in _SatisfiesTypeConstraint
    ", ".join(dtypes.as_dtype(x).name for x in allowed_list)))
```

because the `Inference` base class casts all members of the data dict to float32.

**Comment by jhaberstro** [comment]
For reference, here is a full reproducing test case.

```
import edward as ed
import numpy as np
import tensorflow as tf
from edward.models import Dirichlet, Categorical, PointMass

#
# data setup
#

# define the dataset and a priori choices about the model
vocabulary = ["A", "B", "C", "D"]
documents = ["ABABABCBAB",
             "CDCDCDCDCD",
             "ABABABABAB",
             "CDCDCDCDCD",
             'ABACABABAB']
num_topics = 2
num_documents = len(documents)
num_words = len(vocabulary)

# transform the dataset into a numerical form the model can use
word_to_id = {vocabulary[idx]:idx for idx in range(len(vocabulary))}
id_vocabulary = [word_to_id[word] for word in vocabulary]
id_documents = [[word_to_id[word] for word in doc] for doc in documents]

#
# define the model
#

# betas[k] is a probability distribution over the words for topic k
betas = Dirichlet(alpha=tf.zeros([num_topics, num_words]) + 1.0, validate_args=True)

# thetas[d] is a probability distribution on the topics for document d
thetas = Dirichlet(alpha=tf.zeros([num_documents, num_topics]) + 0.1, validate_args=True)

# c[d][i] is the topic assignment of the ith word in document d
c = [[Categorical(logits=ed.logit(tf.gather(thetas, d)), validate_args=True) for i in range(len(documents[d]))] for d in range(num_documents)]

# x[d][i] represents the ith word from document d such that it is
# drawn from categorical distribution represented by betas[c[d][i]]
batched_logits = [[ed.logit(tf.gather(betas, c[d][i])) for i in range(len(documents[d]))] for d in range(num_documents)]
batched_logits = tf.pack(batched_logits)
x = Categorical(logits=batched_logits, validate_args=True)

#
# perform inference
#

# create the variational parameters
qbetas_gamma = tf.nn.softplus(tf.Variable(tf.random_normal([num_topics, num_words])))
qthetas_alpha = tf.nn.softplus(tf.Variable(tf.random_normal([num_documents, num_topics])))

# create the variational distributions using the variational parameters
qbetas = Dirichlet(alpha=qbetas_gamma, validate_args=True)
qthetas = Dirichlet(alpha=qthetas_alpha, validate_args=True)

inference = ed.KLqp({betas: qbetas, thetas: qthetas},
                    {x: tf.constant(id_documents)})

inference.run()
```

**Comment by dustinvtran** [comment]
thanks for the note. it looks like removing the cast to float32 indeed fixes this. i'd like you to get contributor credit. can you make that change in both places where the casting occurs? i can help make the placeholder's compatible with non-float32 data and also look into any of the examples that fail once the casting is removed.

there's another error after the fix when running the above script. this seems to be caused by dirichlet producing essentially degenerate samples (that are exactly 1); this seems to be a separate issue (see https://github.com/blei-lab/edward/issues/314#issuecomment-271059579).

**Comment by jhaberstro** [comment]
No problem. I'd be happy to make the change and update any examples that may start failing, but I first need to get legal approval  from my workplace before contributing open-source code. I'll start the process for that, but if I'm not approved quickly, I won't be offended if you decide to make the change prior.

As for the degenerate sample dirichlet bug, I was able to work around that by clamping to [0.1, 0.9] the variational parameters and the tensors I pass as logits to `Categorical`, but I still don't manage to learn the topics at all. I imagine a proper fix is needed.

**Comment by dustinvtran** [comment]
both points make sense; thanks for the update.

**Comment by dustinvtran** [comment]
hi @jhaberstro, any progress on getting legal approval?

**Comment by jhaberstro** [comment]
I've put in the request with my management and was told they would start the process immediately, but I don't know how long it will take.

**Comment by dustinvtran** [comment]
fixed in #421

**Comment by jhaberstro** [comment]
Hi Dustin, I just wanted to give you an update. I was not able to get legal approval. I was told perhaps in the future in a limited capacity, but not at this time.

**Comment by dustinvtran** [comment]
that's unfortunate. thanks for trying though. hope you might be able to contribute in the future.
